### PR TITLE
MCOL-4188 Regression fixes for MCOL-641.

### PR DESCRIPTION
**Pull request by tntnatbry**
1. Make PredicateOperator::setOpType() function wide decimal aware.
  2. Added support for wide decimal in jlf_subquery.cpp::getColumnValue()
     used in scalar subqueries.
  3. Fixed the column index used for fetching wide decimal values from a Row
     when a wide decimal field is used in the order by clause.

### CODE REVIEW

**Comment by drrtuy** [comment]
Good findings. Couple notes on the patch. Could you wait till my decimalToString patch will be merged and replace int128_t variables in your patch with TSInt128. I would discourage you from using int128_t directly from now on in favor of TSInt128.

File: dbcon/joblist/jlf_subquery.cpp [hunk 1]
```diff
@@ -90,19 +90,38 @@ void getColumnValue(ConstantColumn** cc, uint64_t i, const Row& row, const strin
...
-            oss << (data / IDB_pow[row.getScale(i)]);
-
-            if (row.getScale(i) > 0)
+            if (row.getColumnWidth(i) == datatypes::MAXDECIMALWIDTH)
             {
-                if (data > 0)
-                    oss << "." << (data % IDB_pow[row.getScale(i)]);
-                else if (data < 0)
-                    oss << "." << (-data % IDB_pow[row.getScale(i)]);
+                int128_t val128;
```
Comment (drrtuy) [inline dbcon/joblist/jlf_subquery.cpp:95]: I would recomend you to take a look at examples in my decimalToString patch to simplify the boilerplate code.

File: utils/windowfunction/idborderby.cpp [hunk 2]
```diff
@@ -172,8 +172,11 @@ int WideDecimalCompare::operator()(IdbCompare* l, Row::Pointer r1, Row::Pointer
     l->row2().setData(r2);
 
     int ret = 0;
-    int128_t v1 = *(l->row1().getBinaryField_offset<int128_t>(keyColumnOffset));
-    int128_t v2 = *(l->row2().getBinaryField_offset<int128_t>(keyColumnOffset));
+
+    int128_t v1, v2;
+    l->row1().getInt128Field(fSpec.fIndex, v1);
```
Comment (drrtuy) [inline utils/windowfunction/idborderby.cpp:177]: I would ask you to solve issue around wrong keyColumnOffset if it is possible b/c it allows us to skip needless dereferencing operation getting the values.

Comment (tntnatbry) [reply utils/windowfunction/idborderby.cpp:177]: I haven't yet addressed this part yet. I have added a comment to MCOL-4188 for this so we don't forget.

**Comment by tntnatbry** [comment]
Some of the comments are addressed in the rebased PR: https://github.com/mariadb-corporation/mariadb-columnstore-engine/pull/1612

File: utils/windowfunction/idborderby.cpp (hunk 2 above)
Comment (drrtuy) [reply utils/windowfunction/idborderby.cpp:177]: OK. Go ahead. I'll handle this.

File: dbcon/execplan/predicateoperator.cpp [hunk 3]
```diff
@@ -216,6 +216,10 @@ void PredicateOperator::setOpType(Type& l, Type& r)
                 // should following the result type that MySQL gives
                 fOperationType = l;
                 fOperationType.scale = (l.scale > r.scale ? l.scale : r.scale);
+                fOperationType.precision = (l.precision > r.precision ?
+                    l.precision : r.precision);
```
Comment (abarkov) [inline dbcon/execplan/predicateoperator.cpp:220]: I'd use std::max() here:

 ` fOperationType.precision = std::max(l.precision, r.precision);`

It's shorter and more readable.

File: dbcon/execplan/predicateoperator.cpp [hunk 4]
```diff
@@ -216,6 +216,10 @@ void PredicateOperator::setOpType(Type& l, Type& r)
...
+                fOperationType.colWidth = (l.colWidth > r.colWidth ?
+                    l.colWidth : r.colWidth);
```
Comment (abarkov) [inline dbcon/execplan/predicateoperator.cpp:222]: Same here:
`fOperationType.colWidth = std::max(l.colWidth, r.colWidth);`

**Review by abarkov — COMMENTED** [review]
The patch looks fine for me. Just a minor suggestion - to use std::max() more actively.

**Pull request closed without merging**